pkg/daemon/daemon.go

```diff
@@ -373,28 +385,84 @@ func (d *daemon) processNetworkGUID(
 	return nil
 }
 
-func syncGUIDPool(smClient plugins.SubnetManagerClient, guidPool guid.Pool) error {
-	usedGuids, err := smClient.ListGuidsInUse()
+func (d *daemon) removeStaleGUID(allocatedGUID, existingPkey string) error {
+	parsedPkey, err := utils.ParsePKey(existingPkey)
+	if err != nil {
+		log.Error().Msgf("failed to parse PKey %s with error: %v", existingPkey, err)
+		return err
+	}
+	guidAddr, err := guid.ParseGUID(allocatedGUID)
+	if err != nil {
+		return fmt.Errorf("failed to parse user allocated guid %s with error: %v", allocatedGUID, err)
+	}
+	allocatedGUIDList := []net.HardwareAddr{guidAddr.HardWareAddress()}
+	// Try to remove pKeys via subnet manager in backoff loop
+	if err = wait.ExponentialBackoff(backoffValues, func() (bool, error) {
+		log.Info().Msgf("removing guids of previous pods from pKey %s"+
+			" with subnet manager %s", existingPkey,
+			d.smClient.Name())
+		if err = d.smClient.RemoveGuidsFromPKey(parsedPkey, allocatedGUIDList); err != nil {
+			log.Warn().Msgf("failed to remove guids of removed pods from pKey %s"+
+				" with subnet manager %s with error: %v", existingPkey,
+				d.smClient.Name(), err)
+			return false, nil //nolint:nilerr // retry pattern for exponential backoff
+		}
+		return true, nil
+	}); err != nil {
+		log.Warn().Msgf("failed to remove guids of removed pods from pKey %s"+
+			" with subnet manager %s", existingPkey, d.smClient.Name())
+		return err
+	}
+
+	if err = d.guidPool.ReleaseGUID(allocatedGUID); err != nil {
+		log.Warn().Msgf("failed to release guid \"%s\" with error: %v", allocatedGUID, err)
+		return err
+	}
+	delete(d.guidPodNetworkMap, allocatedGUID)
+	log.Info().Msgf("successfully released %s from pkey %s", allocatedGUID, existingPkey)
+	return nil
+}
+
+func (d *daemon) syncGUIDPool() error {
+	usedGuids, err := d.smClient.ListGuidsInUse()
 	if err != nil {
 		return err
 	}
 
 	// Reset guid pool with already allocated guids to avoid collisions
-	err = guidPool.Reset(usedGuids)
+	err = d.guidPool.Reset(usedGuids)
 	if err != nil {
 		return err
 	}
+
+	// Remove stale GUIDs that are no longer in use by the subnet manager
+	// This handles cleanup of GUIDs from deleted/finished pods
+	for allocatedGUID, podNetworkID := range d.guidPodNetworkMap {
```

Comment:
At the time when syncGuidPool is called, the `d.guidPodNetworkMap` is empty because it's populated in the `d.initPool` method. Since we run both only once, let's combine them into a single function and properly handle the order of operations